data/src/main/java/org/apache/iceberg/data/BaseWriterFactory.java

```diff
@@ -0,0 +1,220 @@
+  private final Map<String, String> properties;
+  private final Schema dataSchema;
+  private final int[] equalityFieldIds;
+  private final Schema equalityDeleteRowSchema;
+  private final Schema positionDeleteRowSchema;
+
+  protected BaseWriterFactory(Map<String, String> properties, Schema dataSchema,
+                              int[] equalityFieldIds, Schema equalityDeleteRowSchema,
+                              Schema positionDeleteRowSchema) {
+    this.properties = properties;
+    this.dataSchema = dataSchema;
+    this.equalityFieldIds = equalityFieldIds;
+    this.equalityDeleteRowSchema = equalityDeleteRowSchema;
+    this.positionDeleteRowSchema = positionDeleteRowSchema;
+  }
+
+  protected abstract void configureDataWrite(Avro.DataWriteBuilder builder);
+  protected abstract void configureEqualityDelete(Avro.DeleteWriteBuilder builder);
+  protected abstract void configurePositionDelete(Avro.DeleteWriteBuilder builder);
+
+  protected abstract void configureDataWrite(Parquet.DataWriteBuilder builder);
+  protected abstract void configureEqualityDelete(Parquet.DeleteWriteBuilder builder);
+  protected abstract void configurePositionDelete(Parquet.DeleteWriteBuilder builder);
+
+  protected Schema equalityDeleteRowSchema() {
+    return equalityDeleteRowSchema;
+  }
+
+  protected Schema positionDeleteRowSchema() {
+    return positionDeleteRowSchema;
+  }
+
+  @Override
+  public DataWriter<T> newDataWriter(EncryptedOutputFile file, FileFormat format,
+                                     PartitionSpec spec, StructLike partition) {
+    Preconditions.checkArgument(spec != null,
+        "Spec must not be null when creating a data writer");
+    Preconditions.checkArgument(spec.isUnpartitioned() || partition != null,
+        "Partition must not be null for partitioned writes");
+
+    OutputFile outputFile = file.encryptingOutputFile();
+    EncryptionKeyMetadata keyMetadata = file.keyMetadata();
+    MetricsConfig metricsConfig = MetricsConfig.fromProperties(properties);
+
+    try {
+      switch (format) {
+        case AVRO:
+          Avro.DataWriteBuilder avroBuilder = Avro.writeData(outputFile)
+              .setAll(properties)
+              .metricsConfig(metricsConfig)
+              .schema(dataSchema)
+              .spec(spec)
+              .partition(partition)
+              .keyMetadata(keyMetadata)
+              .overwrite();
+
+          configureDataWrite(avroBuilder);
+
+          return avroBuilder.buildDataWriter();
+
+        case PARQUET:
+          Parquet.DataWriteBuilder parquetBuilder = Parquet.writeData(outputFile)
+              .setAll(properties)
+              .metricsConfig(metricsConfig)
+              .schema(dataSchema)
+              .spec(spec)
+              .partition(partition)
+              .keyMetadata(keyMetadata)
+              .overwrite();
+
+          configureDataWrite(parquetBuilder);
+
+          return parquetBuilder.buildDataWriter();
+
+        default:
```

Comment:
I think the ORC data writer is needed. Currently, we did not implement the ORC delete writer, so for the equality delete writer and position delete writer,  we don't support ORC now. But for the append-only case, the ORC data writer is necessary.